DuckDuckGo/AutomationServer.swift

```diff
@@ -0,0 +1,340 @@
+                return
+            default:
+                print("Connection is in state \(connection.state).")
+                return
+            }
+            Logger.automationServer.info("Received request! \(String(describing: content)) \(isComplete) \(String(describing: error))")
+
+            if let error {
+                Logger.automationServer.error("Error: \(error)")
+                return
+            }
+
+            if let content {
+                Logger.automationServer.info("Handling content")
+                Task {
+                    await self.processContentWhenReady(connection: connection, content: content)
+                }
+            }
+
+            if !isComplete {
+                Logger.automationServer.info("Handling not complete")
+                self.receive(from: connection)
+            }
+        }
+    }
+    
+    @MainActor
+    func processContentWhenReady(connection: NWConnection, content: Data) async {
+        // Check if loading
+        while self.main.currentTab?.isLoading ?? false {
+            print("Still loading, waiting...")
+            try? await Task.sleep(nanoseconds: 200_000_000) // 0.2 seconds
+        }
+
+        // Proceed when loading is complete
+        Logger.automationServer.info("Handling content")
+        self.handleConnection(connection, content)
+    }
+    
+    func getQueryStringParameter(url: URLComponents, param: String) -> String? {
+        return url.queryItems?.first(where: { $0.name == param })?.value
+    }
+
+    @MainActor
+    func handleConnection(_ connection: NWConnection, _ content: Data) {
+        Logger.automationServer.info("Handling request!")
+        let stringContent = String(decoding: content, as: UTF8.self)
+        // Log first line of string:
+        if let firstLine = stringContent.components(separatedBy: CharacterSet.newlines).first {
+            Logger.automationServer.info("First line: \(firstLine)")
+        }
+
+        // Get url parameter from path
+        // GET / HTTP/1.1
+        if #available(iOS 16.0, *) {
+            let path = /^(GET|POST) (\/[^ ]*) HTTP/
+            if let match = stringContent.firstMatch(of: path) {
+                Logger.automationServer.info("Path: \(match.2)")
+                // Convert the path into a URL object
+                guard let url = URLComponents(string: String(match.2)) else {
+                    Logger.automationServer.error("Invalid URL: \(match.2)")
+                    return // Or handle the error appropriately
+                }
+                switch url.path {
+                case "/navigate":
+                    self.navigate(on: connection, url: url)
+                case "/execute":
+                    self.execute(on: connection, url: url)
+                case "/getUrl":
+                    self.respond(on: connection, response: self.main.currentUrl() ?? "")
```

Comment:
Since this the only place this is used I'd consider just accessing the webView url directly.

var url = self.main.currentTab?.webView.url?.absoluteString
self.respond(on: connection, response: url ?? "")

Then remove those accessors you added.  Mainly because they will be visible to other code and might cause confusion about what the URL is the code should be using.